File: apps/fsm-core-db/src/fsm-instance-lock.ts
import type { DBDeps } from "./custom-type.ts";

import { FSM_SCHEMA } from "./const.ts";






export async function tryFSMDBLock(
  deps: DBDeps,
  fsmInstanceId: string,
): Promise<boolean> {
  const lockedBy = "some-identifier"; // Replace with actual identifier
  try {
    const LOCK_FSM_INSTANCE_FN = `${FSM_SCHEMA}.lock_fsm_instance`;
    const text = `
      SELECT ${LOCK_FSM_INSTANCE_FN}(
        $1::uuid,
        $2::text
      ) AS locked;
    `;
    const values = [fsmInstanceId, lockedBy];
    const res = await deps.db.query<{ locked: boolean }>(text, values);
    return res.rows?.[0]?.locked === true;
  } catch (err) {
    console.error("Error in tryFSMDBLock:", err);

    throw new Error("Failed to acquire FSM DB lock", { cause: err });
  }
}

export async function releaseFSMDBLock(
  deps: DBDeps,
  fsmInstanceId: string,
): Promise<boolean> {
  try {
    const UNLOCK_FSM_INSTANCE_FN = `${FSM_SCHEMA}.unlock_fsm_instance`;
    const text = `
      SELECT ${UNLOCK_FSM_INSTANCE_FN}(
        $1::uuid
      ) AS unlocked;
    `;
    const res = await deps.db.query<{ unlocked: boolean }>(text, [fsmInstanceId]);
    return res.rows?.[0]?.unlocked === true;
  } catch (err) {
    console.error("Error in releaseFSMDBLock:", err);
    throw new Error("Failed to release FSM DB lock", { cause: err });
  }
}

File: apps/fsm-core-db-py/src/fsm_core_db/fsm_instance_lock.py
import asyncpg

from .constants import FSM_SCHEMA

LOCK_FN = f"{FSM_SCHEMA}.lock_fsm_instance"
UNLOCK_FN = f"{FSM_SCHEMA}.unlock_fsm_instance"
LOCKED_BY = "fsm-core-py"


async def try_fsm_db_lock(pool: asyncpg.Pool, fsm_instance_id: str) -> bool:
    """
    Attempt to acquire an advisory lock on an FSM instance.
    Mirrors tryFSMDBLock() in fsm-instance-lock.ts.
    """
    sql = f"SELECT {LOCK_FN}($1::uuid, $2::text) AS locked;"
    async with pool.acquire() as conn:
        row = await conn.fetchrow(sql, fsm_instance_id, LOCKED_BY)
    return bool(row["locked"]) if row else False


async def release_fsm_db_lock(pool: asyncpg.Pool, fsm_instance_id: str) -> bool:
    """
    Release a previously acquired advisory lock on an FSM instance.
    Mirrors releaseFSMDBLock() in fsm-instance-lock.ts.
    """
    sql = f"SELECT {UNLOCK_FN}($1::uuid) AS unlocked;"
    async with pool.acquire() as conn:
        row = await conn.fetchrow(sql, fsm_instance_id)
    return bool(row["unlocked"]) if row else False
